libs/engine/includes/Events/EventQueue.hpp

```diff
@@ -0,0 +1,184 @@
+/*
+** EPITECH PROJECT, 2023
+** R-Type
+** File description:
+** Event encapsulation
+*/
+
+#ifndef R_TYPE_EVENT_HPP
+#define R_TYPE_EVENT_HPP
+
+#include <deque>
+#include <memory>
+#include <stdexcept>
+
+#include "Entity.hpp"
+#include "IEvent.hpp"
+
+#include <iostream>
+#include <utility>
+
+namespace Event {
+
+class EventQueue {
+public:
+    using ValueType = std::unique_ptr<IEvent>;
+    using Container = std::deque<ValueType>;
+    using Iterator = Container::iterator;
+    using ConstIterator = Container::const_iterator;
+
+public:
+    template <class Event> struct EventsIterator {
+
+    public:
+        explicit EventsIterator(Iterator it, Iterator end)
+            : _it(std::move(it))
+            , _end(std::move(end))
+        {
+            while (_it != _end && not dynamic_cast<Event *>(_it->get())) {
+                _it++;
+            }
+        }
+
+        Event &operator*() const
+        {
+            return dynamic_cast<Event &>(*_it);
+        }
+
+        Event *operator->()
+        {
+            return dynamic_cast<Event *>(_it->get());
+        }
+
+        EventsIterator &operator++()
+        {
+            _it++;
+            while (_it != _end && not dynamic_cast<Event *>(_it->get())) {
+                _it++;
+            }
+            return *this;
+        }
+
+        EventsIterator operator++(int)
+        {
+            EventsIterator tmp = *this;
+
+            ++(*this);
+            return tmp;
+        }
+
+        friend bool operator==(EventsIterator const &a, EventsIterator const &b)
+        {
+            return a._it == b._it;
+        };
+
+        friend bool operator!=(EventsIterator const &a, EventsIterator const &b)
+        {
+            return a._it != b._it;
+        };
+
+    private:
+        Iterator _it;
+        Iterator _end;
+    };
+
+public:
+    template <class Event> EventsIterator<Event> beginIterator()
+    {
+        return EventsIterator<Event>(_events.begin(), _events.end());
+    }
+
+    template <class Event> EventsIterator<Event> endIterator()
+    {
+        return EventsIterator<Event>(_events.end(), _events.end());
+    }
```

Comment:
Au lieu de deux fonctions templatés, il faudrait un type EventsView templaté, qui contient une fonction begin et une fonction end
Comme ce que fait la lib standard: une view que l'on peut ensuite utiliser pour faire des range-based for
```
for (auto &closeEvent : EventsView<Close>(eventsQueue)) {}
```